src/uffd_handler/src/memory_region.rs

```diff
@@ -48,18 +71,49 @@ pub fn create_mem_regions(mappings: Vec<GuestRegionUffdMapping>) -> Vec<MemRegio
     mem_regions
 }
 
-/// Deserialize memory mappings received through UDS from Firecracker.
-pub fn deserialize_mappings(msg: &str, size: usize) -> Result<Vec<GuestRegionUffdMapping>> {
-    let mappings = serde_json::from_str::<Vec<GuestRegionUffdMapping>>(msg)
-        .map_err(Error::DeserializeMemoryMappings)?;
+/// Perform sanity checks for deserialized mappings to ensure they are not corrupted:
+/// - The regions must not exceed memory file end.
+/// - The specified file offset and region size must not overflow when added.
+/// - The region end must not overflow usize.
+/// - The total mappings size must match the memory file size.
+fn mappings_sanity_checks(
+    mappings: &Vec<GuestRegionUffdMapping>,
+    file_size: usize,
+) -> Result<(), MemRegionError> {
+    for region in mappings {
+        if let Some(region_end) = region.offset.checked_add(region.size as u64) {
+            if region_end > file_size as u64 {
+                // The region goes beyond file end.
+                return Err(MemRegionError::MappingPastEof);
+            }
+        } else {
+            // The specified file offset and region size cause overflow when added.
+            return Err(MemRegionError::InvalidOffsetLength);
+        }
+
+        if region
+            .base_host_virt_addr
+            .checked_add(region.size)
+            .is_none()
+        {
+            // The region end overflows usize.
+            return Err(MemRegionError::EndRegionOverflow);
+        }
+    }
+
     let memsize: usize = mappings.iter().map(|r| r.size).sum();
```

Comment:
Could we do the sum of the region sizes in the previous loop such that there is no need to pass through the vector twice? Or, do you know if the compiler generally optimize such things? And third question here (probably the most relevant): how many regions, on average, do we have in a snap file (a low number could be in favor to let it as it is for code visibility)?

Reply:
The regions are the actual guest memory regions, so that will be maximum 2 on x86_64 (because of the MMIO gap) and one on ARM. I *think* the compiler should optimize this, but I am not 100% sure.

Reply:
I'm ok to let as it is in this case. Thanks for clarifications!